### PR TITLE
Add wildcard routes managed by GC Articles

### PR DESCRIPTION
**Pull request by timarney**
Adds a wildcard route handler to serve content that is a valid page from GC Articles

### Updates

1) Add wildcard route handling
```
@main.route("/<path:path>")
def page_content(path=""):
```

2) Moves the "wp nav request" into a standalone function

https://notification-feature-wi-1m1uol.herokuapp.com/wild-card

3) Adds find_item_url(nav_items, path) --- to validate the requested path is a valid page from a GC Articles request.

### Testing

Locally visit http://localhost:6012/wild-card/ --- wildcard route handler should handle and load WP content

Screen Shot 2022-01-12 at 11 57 57 AM

Entering a "Bad Url" i.e. http://localhost:6012/bad_url should still result in a 404 as WordPress doesn't have a menu entry for it.

## Further testing
Create a GC Articles page i.e. https://articles.cdssandbox.xyz/notification-gc-notify/wp-admin/post.php?post=200&action=edit&lang=en

Add a link to the GC Articles menu:
Ensure the url = the slug i.e. "/slug-name-for-page"  
Screen Shot 2022-01-13 at 9 13 49 AM

### CODE REVIEW

File: app/main/views/index.py [hunk 1]
```diff
@@ -23,8 +23,10 @@
 from app.utils import (
     Spreadsheet,
     documentation_url,
+    find_item_url,
     get_latest_stats,
     get_logo_cdn_domain,
+    get_nav_wp,
```
Comment (jimleroyer) [inline app/main/views/index.py:29]: Starting to think we might move away the GCArticles utilities into their module, e.g. `app/articles`, what do you think?

File: app/main/views/index.py [hunk 2]
```diff
@@ -396,3 +361,28 @@ def messages_status():
 @main.route("/templates", endpoint="redirect_format")
 def old_page_redirects():
     return redirect(url_for(request.endpoint.replace("redirect_", "")), code=301)
+
+
+# --- Wildcard for API-driven pages --- #
```
Comment (jimleroyer) [inline app/main/views/index.py:366]:
```suggestion
# --- Dynamic routes handling for GCArticles API-driven pages --- #
```

File: app/utils.py [hunk 3]
```diff
@@ -57,6 +57,7 @@
     "validation-failed",
 ]
 REQUESTED_STATUSES = SENDING_STATUSES + DELIVERED_STATUSES + FAILURE_STATUSES
+CACHE_TTL = 86400
```
Comment (jimleroyer) [inline app/utils.py:60]: Can we rename this to GC_ARTICLE_CACHE_TTL or something similar as this applies only to GCArticles retrieved content?

File: app/utils.py [hunk 4]
```diff
@@ -786,14 +787,36 @@ def request_content(endpoint: str, params={"slug": "", "lang": "en"}) -> Union[l
...
-
         if cache.get(cache_key):
+            current_app.logger.info(f"Cache hit: {cache_key}")
             return cache.get(cache_key)
         else:
             current_app.logger.info(f"Cache miss: {cache_key}")
             return None
+
+
+def find_item_url(items=[], url="") -> Union[dict, None]:
```
Comment (jimleroyer) [inline app/utils.py:802]:
```suggestion
def find_item_url(items=[], url="") -> Optional[dict]:
```

File: app/utils.py [hunk 5]
```diff
@@ -786,14 +787,36 @@ def request_content(endpoint: str, params={"slug": "", "lang": "en"}) -> Union[l
...
+    found = list(filter(lambda item: item["url"] == url, items))
+    if len(found) == 0:
+        return None
+    return found
+
+
+def get_nav_wp(locale) -> Union[list, None]:
```
Comment (jimleroyer) [inline app/utils.py:809]:
```suggestion
def get_nav_wp(locale: str) -> Optional[list]:
```

File: app/main/views/index.py [hunk 6]
```diff
@@ -396,3 +361,28 @@ def messages_status():
...
+@main.route("/<path:path>")
+def page_content(path=""):
+    locale = get_current_locale(current_app)
+    nav_items = get_nav_wp(locale)
```
Comment (jimleroyer) [inline app/main/views/index.py:370]: Hmm would we benefit from caching these items? given how much this will be hit for any dynamic path?

**Review by jimleroyer — COMMENTED** [review]
Very nice work, got a few suggestions, comments and questions.

File: app/main/views/index.py (hunk 1 above)
Comment (pcraig3) [reply app/main/views/index.py]: Done as of: https://github.com/cds-snc/notification-admin/pull/1215/commits/9e799f37a37f53e017c6932ebadf4c1ac8be99ea

**Review by jimleroyer — APPROVED** [review]
:shipit:

**Pull request merged**